src/datasource/gateway.js

```diff
@@ -0,0 +1,273 @@
+      throw {
+        message: error,
+      }
+    }
+
+    throw error;
+
+  }
+
+  perform = async (requestName, ...args) => {
+    const reqConfig = typeof this.requests[requestName] === 'function' ?
+      this.requests[requestName](...args) :
+      this.requests[requestName];
+
+    const response = await this
+      .provider
+      .http
+      .request(reqConfig)
+      .then(this.parseSuccessfulResponse)
+      .catch(this.parseErrorResponse);
+
+    return response;
+  }
+
+  // does not really need to be bound to `this`, but keeping it for semantics' sake.
+  // arguments explicit forward for clear signature lookup, avoid using `...args` 
+  _getHeight = this.perform.bind(this, 'getHeight');
+  _getAccount = (id) => this.perform.call(this, 'getAccount', id);
+  _getTransaction = (id) => this.perform.call(this, 'getTransaction', id);
+  _getBlock = (height) => this.perform.call(this, 'getBlock', height);
+  _getBlockTxs = (height, prove, page, per_page) => this.perform.call(this, 'getBlockTxs', height, prove, page, per_page);
+  _getApps = (staking_status, height, blockchain, page, per_page) => this.perform.call(this, 'getApps', staking_status, height, blockchain, page, per_page);
+  _getSupply = this.perform.bind(this, 'getSupply');
+  _getNodes = (staking_status, jailed_status, height, blockchain, page, per_page) => this.perform.call(this, 'getNodes', staking_status, jailed_status, height, blockchain, page, per_page);
+
+  // For semantic separation, and for "ease of middlewaring" when necessary.
+  query = {
+    getHeight: this._getHeight,
+    getAccount: this._getAccount,
+    getTransaction: this._getTransaction,
+    getBlock: this._getBlock,
+    getBlockTxs: this._getBlockTxs,
+    getApps: this._getApps,
+    getSupply: this._getSupply,
+    getNodes: this._getNodes,
+  }
+}
+
+/**
+ * Exposes registered/allowed gateway queries.
+ * This layer is added for gateway level logic control.
+ */
+class GatewayClient {
+  constructor(httpProvider, requestsController, config) {
+    this.controller = requestsController.use(httpProvider)
+    this.config     = config;
+  }
+
+  queries = [
+    'getHeight',
+    'getAccount',
+    'getTransaction',
+    'getBlock',
+    'getBlockTxs',
+    'getApps',
+    'getSupply',
+    'getNodes',
+  ]
+
+  /**
+   * @returns {BigInt}
+   */
+  async makeQuery(queryName, ...args) {
+    if (!(this.queries.includes(queryName) > -1)) {
+      throw Errors
+    }
+    return await this
+      .controller
+      .query[queryName](...args);
+  }
+
+  getHeight = this.makeQuery.bind(this, 'getHeight');
+  getAccount = this.makeQuery.bind(this, 'getAccount');
+  getTransaction = this.makeQuery.bind(this, 'getTransaction');
+  getBlock = this.makeQuery.bind(this, 'getBlock');
+  getBlockTxs = this.makeQuery.bind(this, 'getBlockTxs');
+  getApps = this.makeQuery.bind(this, 'getApps');
+  getSupply = this.makeQuery.bind(this, 'getSupply');
+  getNodes = this.makeQuery.bind(this, 'getNodes');
```

Comment:
Again, is this really needed? I don't think so

Reply:
The goal in here is to provide a unified API for all queries that's as simple as `gatewayClientInstance.[queryName]`.

To achieve this, we have the simple route of simply implementing the method of the class and getting done with it, which was the initial case (_right?_), then to DRY things up, this piece of code showed up for the DRYable sections of the code.

The non DRYable sections of the code that are custom to their specific use cases are:

* L223: getLatestTransactions
* L230: getTotalStakedApps
* L238: getStakedSupply
* L249: getGreenNodes

So, providing those as direct class methods whilst leaving the others are auxiliaries accessed through `makeQuery` is not super homogeneous.

I don't think DRYing code is up for discussion, what could be up for discussion is, which API do we want to go with (_to achieve in-class implementation homogeny_) is it `GatewayClient.makeQuery` and make the Non-DRYable ones accessed through this method.

Reply:
Hmm, I'd say this argument makes sense. For now I wouldn't debate over the API too much, so I'd go with the first option (`.makeQuery` and just have the other as edge cases) for the sake of being nimble.

Reply:
👍🏻